File: tools/saves/ctsave.py
"""Stream model of the Chrono Trigger (2018 Steam/mobile port) save payload.

See REPORT.md #3 for the format. The payload is a serialized *stream*, not a
fixed C struct: names are length-prefixed ASCII, so every field after the
name table shifts if a name's length changes. CtSave.parse()/.serialize()
therefore walk the stream sequentially in both directions.

Container framing (REPORT.md #3.1): on disk a save is
    8-byte header (IV xor magic) + Blowfish-CBC ciphertext
which decrypts (via ctcrypto.py) to
    payload[N] + u16 checksum + pad-to-8-byte-boundary + u32 LE N
`load()`/`save()` handle that trailer; CtSave itself only models payload[N].
"""
import struct

import ctcrypto

CHAR_SIZE = 0x58
NUM_CHARS = 7
NUM_NAMES = 10
REGION_A_SIZE = 0x200
FLAGS_SIZE = 0x200
BLOCK_400_SIZE = 0x30

# (capacity, category byte) for the six inventory sections, in stream order.
INVENTORY_SECTIONS = [
    ("weapons", 111, 0x00),
    ("armour", 50, 0x10),
    ("helmets", 39, 0x20),
    ("accessories", 59, 0x30),
    ("consumables", 43, 0x40),
    ("key_items", 45, 0x50),
]
TECH_BLOCK_SIZE = 45


class CtChar:
    """One 0x58-byte character record (REPORT.md #4, port column).

    Backed by a mutable bytearray (`raw`); every named field is a
    property that reads/writes directly into `raw`, so `bytes(self.raw)`
    is always an exact, up-to-date serialization -- there is no separate
    field state to keep in sync.
    """

    def __init__(self, raw: bytes):
        assert len(raw) == CHAR_SIZE
        self.raw = bytearray(raw)

    def _u8(self, off, value=None):
        if value is None:
            return self.raw[off]
        self.raw[off] = value & 0xFF

    def _u16(self, off, value=None):
        if value is None:
            return struct.unpack_from('<H', self.raw, off)[0]
        struct.pack_into('<H', self.raw, off, value & 0xFFFF)

    def _u32(self, off, value=None):
        if value is None:
            return struct.unpack_from('<I', self.raw, off)[0]
        struct.pack_into('<I', self.raw, off, value & 0xFFFFFFFF)

    # +00..+03
    id = property(lambda s: s._u8(0x00), lambda s, v: s._u8(0x00, v))
    const_01 = property(lambda s: s._u8(0x01), lambda s, v: s._u8(0x01, v))
    const_02 = property(lambda s: s._u8(0x02), lambda s, v: s._u8(0x02, v))
    const_03 = property(lambda s: s._u8(0x03), lambda s, v: s._u8(0x03, v))

    # HP/MP -- note the port swaps max/current order vs. the SNES record.
    max_hp = property(lambda s: s._u16(0x04), lambda s, v: s._u16(0x04, v))
    cur_hp = property(lambda s: s._u16(0x06), lambda s, v: s._u16(0x06, v))
    max_mp = property(lambda s: s._u16(0x08), lambda s, v: s._u16(0x08, v))
    cur_mp = property(lambda s: s._u16(0x0A), lambda s, v: s._u16(0x0A, v))
    base_max_hp = property(lambda s: s._u16(0x0C), lambda s, v: s._u16(0x0C, v))

    base_power = property(lambda s: s._u8(0x0E), lambda s, v: s._u8(0x0E, v))
    base_stamina = property(lambda s: s._u8(0x0F), lambda s, v: s._u8(0x0F, v))
    base_speed = property(lambda s: s._u8(0x10), lambda s, v: s._u8(0x10, v))
    base_magic = property(lambda s: s._u8(0x11), lambda s, v: s._u8(0x11, v))
    base_hit = property(lambda s: s._u8(0x12), lambda s, v: s._u8(0x12, v))
    base_evade = property(lambda s: s._u8(0x13), lambda s, v: s._u8(0x13, v))
    base_mdef = property(lambda s: s._u8(0x14), lambda s, v: s._u8(0x14, v))
    level = property(lambda s: s._u8(0x15), lambda s, v: s._u8(0x15, v))

    exp = property(lambda s: s._u32(0x16), lambda s, v: s._u32(0x16, v))
    tp_next = property(lambda s: s._u32(0x1A), lambda s, v: s._u32(0x1A, v))
    # +0x1E..+0x2B (14 bytes): zeros in every observed fixture.

    equip_weapon = property(lambda s: s._u16(0x2C), lambda s, v: s._u16(0x2C, v))
    equip_armor = property(lambda s: s._u16(0x2E), lambda s, v: s._u16(0x2E, v))
    equip_helmet = property(lambda s: s._u16(0x30), lambda s, v: s._u16(0x30, v))
    equip_accessory = property(lambda s: s._u16(0x32), lambda s, v: s._u16(0x32, v))

    exp_next = property(lambda s: s._u16(0x34), lambda s, v: s._u16(0x34, v))
    tp_related = property(lambda s: s._u16(0x36), lambda s, v: s._u16(0x36, v))

    @property
    def growth(self):
        return bytes(self.raw[0x38:0x3E])

    @growth.setter
    def growth(self, value):
        assert len(value) == 6
        self.raw[0x38:0x3E] = value

    @property
    def cur_stats(self):
        """9 bytes: power, stamina, speed, magic, hit, evade, mdef, attack, defence."""
        return bytes(self.raw[0x3E:0x47])

    @cur_stats.setter
    def cur_stats(self, value):
        assert len(value) == 9
        self.raw[0x3E:0x47] = value

    @property
    def per_char_constants(self):
        return bytes(self.raw[0x47:0x4B])

    @per_char_constants.setter
    def per_char_constants(self, value):
        assert len(value) == 4
        self.raw[0x47:0x4B] = value

    # +0x4B..+0x57 (13 bytes): zeros in every observed fixture.

    def serialize(self) -> bytes:
        return bytes(self.raw)


def _read_pascal_str(data: bytes, off: int):
    """One length-prefixed name: u8 byte-length, then that many bytes.

    REPORT.md #3.2 describes these as ASCII, and they are for every fixture
    except one: X12_dreamsepilogue.bin has a custom Chinese name for Epoch,
    stored as UTF-8 (length byte 0x0c = 12 bytes for 4 characters). Decode as
    UTF-8 (a superset of ASCII) rather than pure ASCII to handle both.
    """
    n = data[off]
    s = data[off + 1: off + 1 + n].decode('utf-8')
    return s, off + 1 + n


def _write_pascal_str(s: str) -> bytes:
    b = s.encode('utf-8')
    assert len(b) <= 255
    return bytes([len(b)]) + b


class CtSave:
    def __init__(self):
        self.region_a = bytearray(REGION_A_SIZE)
        self.flags = bytearray(FLAGS_SIZE)
        self.block_400 = bytearray(BLOCK_400_SIZE)
        self.chars = []  # 7 CtChar
        self.inventory = {}  # section name -> list[(idx, count)], fixed capacity
        self.tech_block = bytearray(TECH_BLOCK_SIZE)
        self.names = []  # 10 strings
        self.party = b'\x00\x00\x00'
        self.reserve = b'\x00' * 6
        self.recruited_mask = 0
        self.byte_0x0A = 0
        self.gold = 0
        self.counter1 = 0
        self.play_time_seconds = 0
        self.counter2 = 0
        self.zero_u16 = 0
        self.location_name_id = 0
        self.era_mask = 0
        self.tail = b''  # opaque: everything from here to the end of payload[N]
        self.checksum = 0  # trailer u16; unknown algorithm, see load()/save()

    @staticmethod
    def parse(payload: bytes) -> "CtSave":
        s = CtSave()
        off = 0
        s.region_a = bytearray(payload[off:off + REGION_A_SIZE]); off += REGION_A_SIZE
        s.flags = bytearray(payload[off:off + FLAGS_SIZE]); off += FLAGS_SIZE
        s.block_400 = bytearray(payload[off:off + BLOCK_400_SIZE]); off += BLOCK_400_SIZE

        s.chars = []
        for _ in range(NUM_CHARS):
            s.chars.append(CtChar(payload[off:off + CHAR_SIZE]))
            off += CHAR_SIZE

        s.inventory = {}
        for name, capacity, cat_byte in INVENTORY_SECTIONS:
            entries = []
            for _ in range(capacity):
                idx, cat, count = payload[off], payload[off + 1], payload[off + 2]
                off += 3
                if cat not in (0, cat_byte):
                    raise ValueError(
                        f"unexpected category byte {cat:#04x} in section {name} "
                        f"(expected 0x00 or {cat_byte:#04x})")
                entries.append((idx, count))
            s.inventory[name] = entries

        s.tech_block = bytearray(payload[off:off + TECH_BLOCK_SIZE]); off += TECH_BLOCK_SIZE

        s.names = []
        for _ in range(NUM_NAMES):
            name, off = _read_pascal_str(payload, off)
            s.names.append(name)

        s.party = payload[off:off + 3]; off += 3
        s.reserve = payload[off:off + 6]; off += 6
        s.recruited_mask = payload[off]; off += 1
        s.byte_0x0A = payload[off]; off += 1
        s.gold = payload[off] | (payload[off + 1] << 8) | (payload[off + 2] << 16); off += 3
        s.counter1 = payload[off]; off += 1
        s.play_time_seconds = (payload[off] | (payload[off + 1] << 8)
                                | (payload[off + 2] << 16)); off += 3
        s.counter2 = payload[off]; off += 1
        s.zero_u16 = struct.unpack_from('<H', payload, off)[0]; off += 2
        s.location_name_id = struct.unpack_from('<H', payload, off)[0]; off += 2
        s.era_mask = struct.unpack_from('<H', payload, off)[0]; off += 2

        s.tail = payload[off:]
        return s

    def serialize(self) -> bytes:
        out = bytearray()
        out += self.region_a
        out += self.flags
        out += self.block_400
        for c in self.chars:
            out += c.serialize()

        for name, capacity, cat_byte in INVENTORY_SECTIONS:
            entries = self.inventory[name]
            assert len(entries) == capacity, f"{name}: expected {capacity} entries, got {len(entries)}"
            for idx, count in entries:
                cat = cat_byte if (idx or count) else 0
                out += bytes([idx & 0xFF, cat, count & 0xFF])

        out += self.tech_block

        for name in self.names:
            out += _write_pascal_str(name)

        out += self.party
        out += self.reserve
        out.append(self.recruited_mask)
        out.append(self.byte_0x0A)
        out += bytes([self.gold & 0xFF, (self.gold >> 8) & 0xFF, (self.gold >> 16) & 0xFF])
        out.append(self.counter1)
        t = self.play_time_seconds
        out += bytes([t & 0xFF, (t >> 8) & 0xFF, (t >> 16) & 0xFF])
        out.append(self.counter2)
        out += struct.pack('<H', self.zero_u16)
        out += struct.pack('<H', self.location_name_id)
        out += struct.pack('<H', self.era_mask)
        out += self.tail
        return bytes(out)


def load(path: str) -> CtSave:
    with open(path, 'rb') as f:
        data = f.read()
    plain = ctcrypto.decrypt(data)
    n = struct.unpack_from('<I', plain, len(plain) - 4)[0]
    payload = plain[:n]
    save = CtSave.parse(payload)
    save.checksum = struct.unpack_from('<H', plain, n)[0]
    return save


def save(path: str, ct_save: CtSave, iv: bytes = None):
    """Serialize and encrypt `ct_save` to `path`.

    The trailer checksum algorithm is unknown (REPORT.md #3.1) -- this writes
    whatever is currently in `ct_save.checksum` (default 0, or whatever was
    loaded from an existing file via `load()`). Pad bytes are zero-filled
    (the original format's pad bytes are documented as uninitialised
    garbage, so zero is as good a choice as any).
    """
    payload = ct_save.serialize()
    n = len(payload)
    trailer = bytearray()
    trailer += struct.pack('<H', ct_save.checksum & 0xFFFF)
    total = n + len(trailer) + 4
    pad = (-total) % 8
    trailer += bytes(pad)
    trailer += struct.pack('<I', n)
    plain = payload + bytes(trailer)
    if iv is None:
        iv = bytes(8)
    data = ctcrypto.encrypt(plain, iv=iv)
    with open(path, 'wb') as f:
        f.write(data)

File: app/src/main/java/com/kalenjohnson/chronoduo/saveimport/CtSave.java
package com.kalenjohnson.chronoduo.saveimport;

import java.io.ByteArrayOutputStream;
import java.io.UnsupportedEncodingException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Stream model of the Chrono Trigger (2018 Steam/mobile port) save payload,
 * ported byte-for-byte from tools/saves/ctsave.py. See REPORT.md #3 for the
 * format. The payload is a serialized *stream*, not a fixed struct: names
 * are length-prefixed, so every field after the name table shifts if a
 * name's length changes -- {@link #parse} and {@link #serialize} walk the
 * stream sequentially in both directions, same as the Python.
 */
public final class CtSave {
    public static final int CHAR_SIZE = 0x58;
    public static final int NUM_CHARS = 7;
    public static final int NUM_NAMES = 10;
    public static final int REGION_A_SIZE = 0x200;
    public static final int FLAGS_SIZE = 0x200;
    public static final int BLOCK_400_SIZE = 0x30;
    public static final int TECH_BLOCK_SIZE = 45;

    /** (section name, capacity, category byte), in stream order. */
    public static final class InventorySection {
        public final String name;
        public final int capacity;
        public final int categoryByte;

        public InventorySection(String name, int capacity, int categoryByte) {
            this.name = name;
            this.capacity = capacity;
            this.categoryByte = categoryByte;
        }
    }

    public static final InventorySection[] INVENTORY_SECTIONS = {
            new InventorySection("weapons", 111, 0x00),
            new InventorySection("armour", 50, 0x10),
            new InventorySection("helmets", 39, 0x20),
            new InventorySection("accessories", 59, 0x30),
            new InventorySection("consumables", 43, 0x40),
            new InventorySection("key_items", 45, 0x50),
    };

    /** One (idx, count) inventory entry. */
    public static final class Entry {
        public int idx, count;

        public Entry(int idx, int count) {
            this.idx = idx;
            this.count = count;
        }
    }

    /**
     * One 0x58-byte character record (REPORT.md #4, port column). Backed by
     * a mutable byte[] (`raw`); every field is a get/set pair reading/
     * writing directly into `raw`, mirroring ctsave.py's property-based
     * CtChar so {@link #serialize} is always exact.
     */
    public static final class CtChar {
        public final byte[] raw;

        public CtChar(byte[] raw) {
            if (raw.length != CHAR_SIZE) throw new IllegalArgumentException("bad char record size");
            this.raw = raw.clone();
        }

        private int u8(int off) {
            return raw[off] & 0xFF;
        }

        private void u8(int off, int v) {
            raw[off] = (byte) (v & 0xFF);
        }

        private int u16(int off) {
            return (raw[off] & 0xFF) | ((raw[off + 1] & 0xFF) << 8);
        }

        private void u16(int off, int v) {
            raw[off] = (byte) (v & 0xFF);
            raw[off + 1] = (byte) ((v >>> 8) & 0xFF);
        }

        private long u32(int off) {
            return (raw[off] & 0xFFL) | ((raw[off + 1] & 0xFFL) << 8)
                    | ((raw[off + 2] & 0xFFL) << 16) | ((raw[off + 3] & 0xFFL) << 24);
        }

        private void u32(int off, long v) {
            raw[off] = (byte) (v & 0xFF);
            raw[off + 1] = (byte) ((v >>> 8) & 0xFF);
            raw[off + 2] = (byte) ((v >>> 16) & 0xFF);
            raw[off + 3] = (byte) ((v >>> 24) & 0xFF);
        }

        public int getId() { return u8(0x00); }
        public void setId(int v) { u8(0x00, v); }
        public int getConst01() { return u8(0x01); }
        public void setConst01(int v) { u8(0x01, v); }
        public int getConst02() { return u8(0x02); }
        public void setConst02(int v) { u8(0x02, v); }
        public int getConst03() { return u8(0x03); }
        public void setConst03(int v) { u8(0x03, v); }

        // HP/MP -- the port swaps max/current order vs. the SNES record.
        public int getMaxHp() { return u16(0x04); }
        public void setMaxHp(int v) { u16(0x04, v); }
        public int getCurHp() { return u16(0x06); }
        public void setCurHp(int v) { u16(0x06, v); }
        public int getMaxMp() { return u16(0x08); }
        public void setMaxMp(int v) { u16(0x08, v); }
        public int getCurMp() { return u16(0x0A); }
        public void setCurMp(int v) { u16(0x0A, v); }
        public int getBaseMaxHp() { return u16(0x0C); }
        public void setBaseMaxHp(int v) { u16(0x0C, v); }

        public int getBasePower() { return u8(0x0E); }
        public void setBasePower(int v) { u8(0x0E, v); }
        public int getBaseStamina() { return u8(0x0F); }
        public void setBaseStamina(int v) { u8(0x0F, v); }
        public int getBaseSpeed() { return u8(0x10); }
        public void setBaseSpeed(int v) { u8(0x10, v); }
        public int getBaseMagic() { return u8(0x11); }
        public void setBaseMagic(int v) { u8(0x11, v); }
        public int getBaseHit() { return u8(0x12); }
        public void setBaseHit(int v) { u8(0x12, v); }
        public int getBaseEvade() { return u8(0x13); }
        public void setBaseEvade(int v) { u8(0x13, v); }
        public int getBaseMdef() { return u8(0x14); }
        public void setBaseMdef(int v) { u8(0x14, v); }
        public int getLevel() { return u8(0x15); }
        public void setLevel(int v) { u8(0x15, v); }

        public long getExp() { return u32(0x16); }
        public void setExp(long v) { u32(0x16, v); }
        public long getTpNext() { return u32(0x1A); }
        public void setTpNext(long v) { u32(0x1A, v); }
        // +0x1E..+0x2B (14 bytes): zeros in every observed fixture.

        public int getEquipWeapon() { return u16(0x2C); }
        public void setEquipWeapon(int v) { u16(0x2C, v); }
        public int getEquipArmor() { return u16(0x2E); }
        public void setEquipArmor(int v) { u16(0x2E, v); }
        public int getEquipHelmet() { return u16(0x30); }
        public void setEquipHelmet(int v) { u16(0x30, v); }
        public int getEquipAccessory() { return u16(0x32); }
        public void setEquipAccessory(int v) { u16(0x32, v); }

        public int getExpNext() { return u16(0x34); }
        public void setExpNext(int v) { u16(0x34, v); }
        public int getTpRelated() { return u16(0x36); }
        public void setTpRelated(int v) { u16(0x36, v); }

        public byte[] getGrowth() {
            byte[] out = new byte[6];
            System.arraycopy(raw, 0x38, out, 0, 6);
            return out;
        }

        public void setGrowth(byte[] value) {
            if (value.length != 6) throw new IllegalArgumentException("growth must be 6 bytes");
            System.arraycopy(value, 0, raw, 0x38, 6);
        }

        /** 9 bytes: power, stamina, speed, magic, hit, evade, mdef, attack, defence. */
        public byte[] getCurStats() {
            byte[] out = new byte[9];
            System.arraycopy(raw, 0x3E, out, 0, 9);
            return out;
        }

        public void setCurStats(byte[] value) {
            if (value.length != 9) throw new IllegalArgumentException("curStats must be 9 bytes");
            System.arraycopy(value, 0, raw, 0x3E, 9);
        }

        public byte[] getPerCharConstants() {
            byte[] out = new byte[4];
            System.arraycopy(raw, 0x47, out, 0, 4);
            return out;
        }

        public void setPerCharConstants(byte[] value) {
            if (value.length != 4) throw new IllegalArgumentException("perCharConstants must be 4 bytes");
            System.arraycopy(value, 0, raw, 0x47, 4);
        }
        // +0x4B..+0x57 (13 bytes): zeros in every observed fixture.

        public byte[] serialize() {
            return raw.clone();
        }

        CtChar deepCopy() {
            return new CtChar(raw);
        }
    }

    public byte[] regionA = new byte[REGION_A_SIZE];
    public byte[] flags = new byte[FLAGS_SIZE];
    public byte[] block400 = new byte[BLOCK_400_SIZE];
    public CtChar[] chars = new CtChar[0]; // 7 CtChar once parsed
    // section name -> fixed-capacity list of entries, in INVENTORY_SECTIONS order
    public Map<String, List<Entry>> inventory = new LinkedHashMap<>();
    public byte[] techBlock = new byte[TECH_BLOCK_SIZE];
    public List<String> names = new ArrayList<>(); // 10 strings
    public byte[] party = {0, 0, 0};
    public byte[] reserve = new byte[6];
    public int recruitedMask;
    public int byte0x0A;
    public long gold;
    public int counter1;
    public long playTimeSeconds;
    public int counter2;
    public int zeroU16;
    public int locationNameId;
    public int eraMask;
    public byte[] tail = new byte[0]; // opaque: everything from here to the end of payload[N]
    public int checksum; // trailer u16; unknown algorithm, see load()/save() equivalents

    public static CtSave parse(byte[] payload) {
        CtSave s = new CtSave();
        int off = 0;
        s.regionA = slice(payload, off, REGION_A_SIZE); off += REGION_A_SIZE;
        s.flags = slice(payload, off, FLAGS_SIZE); off += FLAGS_SIZE;
        s.block400 = slice(payload, off, BLOCK_400_SIZE); off += BLOCK_400_SIZE;

        s.chars = new CtChar[NUM_CHARS];
        for (int i = 0; i < NUM_CHARS; i++) {
            s.chars[i] = new CtChar(slice(payload, off, CHAR_SIZE));
            off += CHAR_SIZE;
        }

        s.inventory = new LinkedHashMap<>();
        for (InventorySection sec : INVENTORY_SECTIONS) {
            List<Entry> entries = new ArrayList<>(sec.capacity);
            for (int i = 0; i < sec.capacity; i++) {
                int idx = payload[off] & 0xFF;
                int cat = payload[off + 1] & 0xFF;
                int count = payload[off + 2] & 0xFF;
                off += 3;
                if (cat != 0 && cat != sec.categoryByte) {
                    throw new IllegalArgumentException(String.format(
                            "unexpected category byte 0x%02X in section %s (expected 0x00 or 0x%02X)",
                            cat, sec.name, sec.categoryByte));
                }
                entries.add(new Entry(idx, count));
            }
            s.inventory.put(sec.name, entries);
        }

        s.techBlock = slice(payload, off, TECH_BLOCK_SIZE); off += TECH_BLOCK_SIZE;

        s.names = new ArrayList<>(NUM_NAMES);
        for (int i = 0; i < NUM_NAMES; i++) {
            int n = payload[off] & 0xFF;
            String name = new String(payload, off + 1, n, StandardCharsets.UTF_8);
            s.names.add(name);
            off += 1 + n;
        }

        s.party = slice(payload, off, 3); off += 3;
        s.reserve = slice(payload, off, 6); off += 6;
        s.recruitedMask = payload[off] & 0xFF; off += 1;
        s.byte0x0A = payload[off] & 0xFF; off += 1;
        s.gold = (payload[off] & 0xFFL) | ((payload[off + 1] & 0xFFL) << 8) | ((payload[off + 2] & 0xFFL) << 16);
        off += 3;
        s.counter1 = payload[off] & 0xFF; off += 1;
        s.playTimeSeconds = (payload[off] & 0xFFL) | ((payload[off + 1] & 0xFFL) << 8)
                | ((payload[off + 2] & 0xFFL) << 16);
        off += 3;
        s.counter2 = payload[off] & 0xFF; off += 1;
        s.zeroU16 = ((payload[off] & 0xFF) | ((payload[off + 1] & 0xFF) << 8)); off += 2;
        s.locationNameId = ((payload[off] & 0xFF) | ((payload[off + 1] & 0xFF) << 8)); off += 2;
        s.eraMask = ((payload[off] & 0xFF) | ((payload[off + 1] & 0xFF) << 8)); off += 2;

        s.tail = slice(payload, off, payload.length - off);
        return s;
    }

    public byte[] serialize() {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        writeBytes(out, regionA);
        writeBytes(out, flags);
        writeBytes(out, block400);
        for (CtChar c : chars) writeBytes(out, c.serialize());

        for (InventorySection sec : INVENTORY_SECTIONS) {
            List<Entry> entries = inventory.get(sec.name);
            if (entries == null || entries.size() != sec.capacity) {
                throw new IllegalStateException(sec.name + ": expected " + sec.capacity + " entries, got "
                        + (entries == null ? 0 : entries.size()));
            }
            for (Entry e : entries) {
                int cat = (e.idx != 0 || e.count != 0) ? sec.categoryByte : 0;
                out.write(e.idx & 0xFF);
                out.write(cat);
                out.write(e.count & 0xFF);
            }
        }

        writeBytes(out, techBlock);

        for (String name : names) {
            byte[] b = utf8(name);
            if (b.length > 255) throw new IllegalStateException("name too long: " + name);
            out.write(b.length);
            writeBytes(out, b);
        }

        writeBytes(out, party);
        writeBytes(out, reserve);
        out.write(recruitedMask & 0xFF);
        out.write(byte0x0A & 0xFF);
        out.write((int) (gold & 0xFF));
        out.write((int) ((gold >>> 8) & 0xFF));
        out.write((int) ((gold >>> 16) & 0xFF));
        out.write(counter1 & 0xFF);
        out.write((int) (playTimeSeconds & 0xFF));
        out.write((int) ((playTimeSeconds >>> 8) & 0xFF));
        out.write((int) ((playTimeSeconds >>> 16) & 0xFF));
        out.write(counter2 & 0xFF);
        out.write(zeroU16 & 0xFF);
        out.write((zeroU16 >>> 8) & 0xFF);
        out.write(locationNameId & 0xFF);
        out.write((locationNameId >>> 8) & 0xFF);
        out.write(eraMask & 0xFF);
        out.write((eraMask >>> 8) & 0xFF);
        writeBytes(out, tail);

        return out.toByteArray();
    }

    /** Deep copy via round-trip, mirroring convert.snes_to_ct's `CtSave.parse(template.serialize())`. */
    public CtSave deepCopy() {
        return parse(serialize());
    }

    private static byte[] utf8(String s) {
        try {
            return s.getBytes("UTF-8");
        } catch (UnsupportedEncodingException e) {
            throw new AssertionError(e);
        }
    }

    private static byte[] slice(byte[] a, int off, int len) {
        byte[] out = new byte[len];
        System.arraycopy(a, off, out, 0, len);
        return out;
    }

    private static void writeBytes(ByteArrayOutputStream out, byte[] b) {
        out.write(b, 0, b.length);
    }
}
